consensus/propagation/have_wants.go

```diff
@@ -85,118 +90,117 @@ func perPeerConcurrentRequestLimit() int64 {
+			if canSend <= 0 {
+				// should never be below zero
 				continue
 			}
 
-			// don't request the part from this peer if we've already requested it
-			// from them.
-			// this check will be helpful once we start requesting parts multiple times.
-			for _, p := range reqs {
-				// p == peer means we have already requested the part from this peer.
-				if p == ps.peer.ID() {
+			var (
+				wants    *proptypes.WantParts
+				parts    *proptypes.CombinedPartSet
+				fullReqs *bits.BitArray
+			)
+			for i := canSend; i > 0; {
+				if len(ps.receivedHaves) == 0 {
+					break
+				}
+
+				have, ok := <-ps.receivedHaves
+				if !ok {
+					return
+				}
+
+				if !blockProp.relevantHave(have.height, have.round) {
 					continue
 				}
-			}
 
-			if want == nil {
-				want = &proptypes.WantParts{
-					Height: req.height,
-					Round:  req.round,
-					Parts:  bits.NewBitArray(int(parts.Total())),
+				if parts == nil {
+					var has bool
+					_, parts, fullReqs, has = blockProp.getAllState(have.height, have.round, false)
+					if !has {
+						blockProp.Logger.Error("couldn't find proposal when filtering requests", "height", have.height, "round", have.round)
+						break
+					}
+				}
+
+				// don't request a part that is already downloaded
+				if parts.BitArray().GetIndex(int(have.index)) {
+					continue
+				}
+
+				// don't request a part that has already hit the request limit
+				if fullReqs.GetIndex(int(have.index)) {
+					continue
+				}
+
+				reqs := blockProp.countRequests(have.height, have.round, int(have.index))
+				if len(reqs) >= ReqLimit() {
+					fullReqs.SetIndex(int(have.index), true)
+					continue
+				}
+
+				// don't request the part from this peer if we've already requested it
+				// from them.
+				for _, p := range reqs {
+					// p == peer means we have already requested the part from this peer.
+					if p == ps.peer.ID() {
+						continue
+					}
+				}
+
+				if wants == nil {
+					wants = &proptypes.WantParts{
+						Height: have.height,
+						Round:  have.round,
+						Parts:  bits.NewBitArray(int(parts.Total())),
+					}
 				}
-			}
-			want.Parts.SetIndex(int(req.index), true)
-			batchRequestCount++
 
-			if len(ps.requestChan) == 0 {
-				err := blockProp.sendWantsThenBroadcastHaves(ps, want)
+				wants.Parts.SetIndex(int(have.index), true)
+
+				err := blockProp.sendWantsThenBroadcastHaves(ps, wants)
 				if err != nil {
 					blockProp.Logger.Error("error sending wants", "err", err)
 				}
-				want = nil
-				batchRequestCount = 0
 			}
 		}
 	}
 }
 
-func (blockProp *Reactor) handleBatchLimit(ps *PeerState, want *proptypes.WantParts) error {
-	if want != nil && !want.Parts.IsEmpty() {
-		err := blockProp.sendWantsThenBroadcastHaves(ps, want)
-		if err != nil {
-			return err
-		}
-	}
-
-	select {
-	case <-blockProp.ctx.Done():
-		return nil
-	case _, ok := <-ps.receivedPart:
-		if !ok {
-			return nil
-		}
-		ps.DecreaseRequestCount(1)
-	}
-	return nil
-}
```

Comment:
we always fill haves before we send anything, so no need to keep track of batch size etc

Reply:
to retro this function, its doing too many things

its meant to block and batch things